### PR TITLE
[Improvement] Automatic selectboxes for choosing schema/version

### PR DESCRIPTION
**Pull request by mproch**
Peek 2020-06-18 16-50

### CODE REVIEW

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala [hunk 1]
```diff
@@ -38,6 +40,18 @@ class CachedConfluentSchemaRegistryClient(val client: CSchemaRegistryClient, sch
       }
     }
 
+
+  override def getAllTopics: Validated[SchemaRegistryError, List[String]] = handleClientError {
+    val pattern = "(.*)-value".r
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala:45]: Nice trick ;)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala [hunk 2]
```diff
@@ -32,4 +32,10 @@ trait SchemaRegistryClient extends Serializable {
     version
       .map(ver => getBySubjectAndVersion(subject, ver))
       .getOrElse(getLatestFreshSchema(subject))
+
+  def getAllTopics: Validated[SchemaRegistryError, List[String]]
+
+  def getAllVersions(schema: String): Validated[SchemaRegistryError, List[Integer]]
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:38]: Not schema but topic.

File: engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/transformer/DynamicParametersSample.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package pl.touk.nussknacker.engine.management.sample.transformer
```
Comment (lciolecki) [inline engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/transformer/DynamicParametersSample.scala:1]: Hmm, why sample.. I think we should do Two traits: DynamicParametersMixin and DynamicParametersSample. After that DynamicParametersSource should extend by DynamicParametersMixin

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala [hunk 4]
```diff
@@ -73,16 +77,53 @@ class CustomNodeValidator(modelData: ModelData) extends NodeDataValidator[Custom
...
-          case Invalid(e) => ValidationPerformed(e.toList, None)
-        }
+        validate(a, nodeData, validationContext, nodeData.outputVar)
+      //TODO: handle 'standard' case
+      case _ => ValidationNotPerformed
+    }
+  }
+}
+
+class SinkNodeValidator(val modelData: ModelData) extends WithParametersNodeValidator[Sink] {
```
Comment (lciolecki) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala:106]: Hmm SinkNodeValidator and SourceNodeValidator are mostly the same.. Maybe we can do some flag to use:
```
modelData.processWithObjectsDefinition.sinkFactories(nodeData.ref.typ)._1.obj or modelData.processWithObjectsDefinition.sourceFactories(nodeData.ref.typ).obj

**Comment by lciolecki** [comment]
Very "guda" job ;) But I think about topic as a select.. When we have many of schemas then can be some problem.. Mayby better way is create autocomplete at FE for it?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala (hunk 4 above)
Comment (mproch) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodevalidation/NodeDataValidator.scala:106]: Yes, but it's on purpose. The difference will emerge when we fix:
//TODO: handle 'standard' case

File: engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/transformer/DynamicParametersSample.scala (hunk 3 above)
Comment (mproch) [reply engine/flink/management/sample/src/main/scala/pl/touk/nussknacker/engine/management/sample/transformer/DynamicParametersSample.scala]: Ok, changed name in previous PR

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala (hunk 2 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala]: :+1: changed

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala (hunk 1 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/client/CachedConfluentSchemaRegistryClient.scala]: "when you have a problem and solve it with regexp..." ;)

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala [hunk 5]
```diff
@@ -34,4 +34,10 @@ trait SchemaRegistryClient extends Serializable {
     version
       .map(ver => getBySubjectAndVersion(topic, ver, isKey))
       .getOrElse(getLatestFreshSchema(topic, isKey))
+
+  def getAllTopics: Validated[SchemaRegistryError, List[String]]
+
+  def getAllVersions(topic: String): Validated[SchemaRegistryError, List[Integer]]
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryClient.scala:40]: I think we should add there isKey: Boolean..

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/LegacyKafkaAvroSourceFactory.scala [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.NodeId
+import pl.touk.nussknacker.engine.api.editor.{DualEditor, DualEditorMode, SimpleEditor, SimpleEditorType}
+import pl.touk.nussknacker.engine.api.process.{ProcessObjectDependencies, Source, TestDataGenerator}
+import pl.touk.nussknacker.engine.api.typed.ReturningType
+import pl.touk.nussknacker.engine.api.{MetaData, MethodToInvoke, ParamName}
+import pl.touk.nussknacker.engine.avro.schemaregistry.{SchemaRegistryKafkaAvroProvider, SchemaRegistryProvider}
+import pl.touk.nussknacker.engine.kafka.KafkaConfig
+import pl.touk.nussknacker.engine.avro.KafkaAvroFactory._
+
+class LegacyKafkaAvroSourceFactory[T: TypeInformation](schemaRegistryProvider: SchemaRegistryProvider[T],
```
Comment (lciolecki) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/LegacyKafkaAvroSourceFactory.scala:16]: Mayby we can mark it as depreicated?

File: engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala [hunk 7]
```diff
@@ -24,7 +24,6 @@ case class FullNameNotCompatible(var first: CharSequence, var middle: CharSequen
       case 0 => first = value.asInstanceOf[CharSequence]
       case 1 => middle = value.asInstanceOf[CharSequence]
       case 2 => last = value.asInstanceOf[CharSequence]
-      case 2 => sex = value.asInstanceOf[CharSequence]
```
Comment (lciolecki) [inline engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: We you removed it?

Comment (mproch) [reply engine/flink/avro-util/src/test/scala/pl/touk/nussknacker/engine/avro/schema/FullNameNotCompatible.scala:27]: addressed in previous PR

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/LegacyKafkaAvroSourceFactory.scala (hunk 6 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/LegacyKafkaAvroSourceFactory.scala:17]: done

**Pull request merged**